brainiak/isfc.py

```diff
@@ -1,226 +1,1104 @@
+
+    summary_statistic : numpy function, default:np.median
+        Summary statistic, either np.median (default) or np.mean
+        
+    n_shifts : int, default:1000
+        Number of randomly shifted samples
+        
+    return_distribution : bool, default:False
+        Optionally return the bootstrap distribution of summary statistics
+        
+    random_state = int, None, or np.random.RandomState, default:None
+        Initial random seed
+
+    Returns
+    -------
+    observed : float, observed ISC (without time-shifting)
+        Actual ISCs
+
+    p : float, p-value
+        p-value based on time-shifting randomization test
+        
+    distribution : ndarray, time-shifts by voxels (optional)
+        Time-shifted null distribution if return_bootstrap=True
+    """
+
+    # Convert list input to 3d and check shapes
+    if type(data) == list:
+        data_shape = data[0].shape
+        for i, d in enumerate(data):
+            if d.shape != data_shape:
+                raise ValueError("All ndarrays in input list "
+                                 "must be the same shape!")
+            if d.ndim == 1:
+                data[i] = d[:, np.newaxis]
+        data = np.dstack(data)
+
+    # Convert input ndarray to 3d and check shape
+    elif type(data) == np.ndarray:
+        if data.ndim == 2:
+            data = data[:, np.newaxis, :]            
+        elif data.ndim == 3:
+            pass
+        else:
+            raise ValueError("Input ndarray should have 2 "
+                             f"or 3 dimensions (got {data.ndim})!")
+
+    # Infer subjects, TRs, voxels and print for user to check
+    n_subjects = data.shape[2]
+    n_TRs = data.shape[0]
+    n_voxels = data.shape[1]
+    
+    # Get actual observed ISC
+    observed = isc(data, pairwise=pairwise, summary_statistic=summary_statistic)
+    
+    # Iterate through randomized shifts to create null distribution
+    distribution = []
+    for i in np.arange(n_shifts):
+        
+        # Random seed to be deterministically re-randomized at each iteration
+        if isinstance(random_state, np.random.RandomState):
+            prng = random_state
+        else:
+            prng = np.random.RandomState(random_state)
+            
+        # Get randomized phase shifts
+        if data.shape[0] % 2 == 0:
+            # Why are we indexing from 1 not zero here? Vector is n_TRs / -1 long?
+            pos_freq = np.arange(1, data.shape[0] // 2)
+            neg_freq = np.arange(data.shape[0] - 1, data.shape[0] // 2, -1)
+        else:
+            pos_freq = np.arange(1, (data.shape[0] - 1) // 2 + 1)
+            neg_freq = np.arange(data.shape[0] - 1, (data.shape[0] - 1) // 2, -1)
+
+        phase_shifts = prng.rand(len(pos_freq), 1, n_subjects) * 2 * np.math.pi
+        
+        # In pairwise approach, apply all shifts then compute pairwise ISCs
+        if pairwise:
+        
+            # Fast Fourier transform along time dimension of data
```

Comment:
Should we pack the following 3 steps: FFT -> shift -> inv FFT as a function, as it is also used in the leave-one-out procedure?

Reply:
This is a good idea and do-able, but I might leave it as-is for now... It's a bit tricky because in the pairwise approach, your input data includes N subjects and you need a set of N random shifts to apply to all subjects, while in the leave-one-out approach, you're only applying a single phase shift to the test subject (average of N - 1 subjects is not shifted).